docs/design/bucket.md

```diff
@@ -0,0 +1,103 @@
+# Bucket: Task Affinity/Anti-Affinity within Job
+
+[@k82cn](http://github.com/k82cn); April 2, 2019
+
+## Motivation
+
+When high performance workload running on Kubernetes, the performance is greatly affected by pod affinity/anti-affinity, e.g.
+ps/worker mode of Tensorflow. The default scheduler support pod affinity/anti-affinity, but there're some weaknesses:
+
+- The default scheduler will check PodAffinity/PodAntiAffinity against all running pods which impacts schedule performance/throughput
+- The default scheduler only check current pod, the coming pods may not have enough resources to meet affinity
+
+To improve the performance of scheduler and workload, a new algorithm is introduced, named as Bucket. This doc describes the function
+specification of the new algorithm.
+
+## Function Specification
+
+### Scope
+
+* In Scope
+
+  - Provide a way for user to define task affinity/anti-affinity within a job
+  - Support selecting a better node for the tasks based on affinity/anti-affinity
+
+* Out Of Scope
+
+  - Do not support "Network Topology", e.g. switches  
+  
+### API
+
+A new plugin is introduced to avoid Job API changes, named as `task-topology`. The arguments of the plugin are defined
+as follow: 
+
+```yaml
+plugins
+  task-topology: ["--affinity", "[[ps-task, worker-task]]", "--anti-affinity", "[[ps-task]]"]
```

Comment:
Have some trouble understanding the `task-topology` usage and explanation

Reply:
I doubt it is too specific that adding these annotations/labels in plugins. If we need to deal with other annotation/label, we need change it and restart it? 

Need we add a section to tell how to use this feature, using annotation or label?

Reply:
In case someone miss the background of `plugins`, it's a existing feature in [volcano](https://github.com/volcano-sh/volcano/blob/ae619b489407207cc5dea21a76d3caf9ffbafb35/pkg/apis/batch/v1alpha1/job.go#L66)

Reply:
> Need we add a section to tell how to use this feature, using annotation or label?

Yes, we need to define its format for other operators, e.g. kubeflow/tf-operator.